src/simulator/ackermann_model.cpp

```diff
@@ -0,0 +1,99 @@
+#include "simulator/ackermann_model.h"
+#include "shared/util/timer.h"
+#include "shared/math/math_util.h"
+
+using f1tenth_simulator::AckermannCurvatureDriveMsg;
+using math_util::AngleDiff;
+using math_util::AngleMod;
+using std::string;
+using std::isfinite;
+
+namespace ackermann {
+
+CONFIG_FLOAT(cMinTurnR, "min_turn_radius");
+CONFIG_FLOAT(cMaxAccel, "max_accel");
+CONFIG_FLOAT(cMaxSpeed, "max_speed");
+CONFIG_FLOAT(cAngularErrorBias, "angular_error_bias");
+CONFIG_FLOAT(cAngularErrorRate, "angular_error_rate");
+CONFIG_STRING(cDriveTopic, "drive_callback_topic");
+
+
+AckermannModel::AckermannModel(const string& config_file, ros::NodeHandle n) :
+    EntityBase(),
+    angular_error_(0, 1),
+    config_reader_({config_file}){
+  // Use the config reader to initialize the subscriber
+  drive_subscriber_ = n.subscribe(
+      cDriveTopic,
+      1,
+      &AckermannModel::DriveCallback,
+      this);
+}
+
+void AckermannModel::DriveCallback(const AckermannCurvatureDriveMsg& msg) {
+  if (!isfinite(msg.velocity) || !isfinite(msg.curvature)) {
+    printf("Ignoring non-finite drive values: %f %f\n",
+        msg.velocity,
+        msg.curvature);
+    return;
+  }
+  last_cmd_ = msg;
+  t_last_cmd_ = GetMonotonicTime();
+}
+
+float AbsBound(float x, float bound) {
+  if (x > 0.0 && x > bound) {
+    return bound;
+  } else if (x < 0.0 && x < -bound) {
+    return -bound;
+  }
+  return x;
+}
+
+void AckermannModel::Step(const double &dt) {
+  // TODO(jaholtz) For faster than real time simulation we may need
+  // a wallclock invariant method for this.
+  static const double kMaxCommandAge = 0.1;
+  if (GetMonotonicTime() > t_last_cmd_ + kMaxCommandAge) {
+    last_cmd_.velocity = 0;
+  }
+  // Epsilon curvature corresponding to a very large radius of turning.
+  static const float kEpsilonCurvature = 1.0 / 1E3;
+  // Commanded speed bounded to motion limit.
+  const float desired_vel = AbsBound(last_cmd_.velocity, cMaxSpeed);
+  // Maximum magnitude of curvature according to turning limits.
+  const float max_curvature = 1.0 / cMinTurnR;
+  // Commanded curvature bounded to turning limit.
+  const float desired_curvature = AbsBound(last_cmd_.curvature, max_curvature);
+  // Indicates if the command is for linear motion.
+  const bool linear_motion = (fabs(desired_curvature) < kEpsilonCurvature);
+
+  const float dv_max = dt * cMaxAccel;
+  const float bounded_dv = AbsBound(desired_vel - vel_, dv_max);
+  vel_ = vel_ + bounded_dv;
+  const float dist = vel_ * dt;
+
+  float dx = 0;
```

Comment:
Why are there manually carried around coordinates here instead of `Eigen::Vector2f`? Using Eigen makes your code shorter, more readable, and faster.